File: sdks/java/src/main/java/com/cavs/sdk/nativebridge/NativeBridge.java
package com.cavs.sdk.nativebridge;

import java.util.function.Consumer;

/**
 * The seam between the Java SDK and the native CAVS library. Handles are
 * opaque {@code long} addresses. The shipped implementation is
 * {@link FfmNativeBridge} (Java Foreign Function &amp; Memory API, JEP 454); a
 * JNA-based backend for the Java 17 baseline can be added behind this same
 * interface without touching {@code CavsClient}.
 */
public interface NativeBridge {

    String version();

    String abiVersion();

    String capabilitiesJson();

    long contextNew();

    void contextFree(long ctx);

    /** Register (or clear, with {@code null}) the progress sink for a context. */
    void setProgressCallback(long ctx, Consumer<String> sink);

    /** Run synchronously; returns the response envelope JSON. */
    String executeSync(long ctx, String operation, String requestJson);

    /** Start a background job; returns its handle (0 on rejection). */
    long startJob(long ctx, String operation, String requestJson);

    /** Poll a job: the envelope JSON if finished, else {@code null}. */
    String pollJob(long job);

    void cancelJob(long job);

    void freeJob(long job);
}

File: sdks/kotlin/src/main/kotlin/com/cavs/sdk/nativebridge/NativeBridge.kt
package com.cavs.sdk.nativebridge

/**
 * The seam between the Kotlin SDK and the native CAVS library. Handles are
 * opaque `Long` addresses. The shipped implementation is [FfmNativeBridge]
 * (Java Foreign Function & Memory API, JEP 454); an alternative backend can
 * slot in behind this interface without touching `CavsClient`.
 */
interface NativeBridge {

    fun version(): String

    fun abiVersion(): String

    fun capabilitiesJson(): String

    fun contextNew(): Long

    fun contextFree(ctx: Long)

    /** Register (or clear, with `null`) the progress sink for a context. */
    fun setProgressCallback(ctx: Long, sink: ((String) -> Unit)?)

    /** Run synchronously; returns the response envelope JSON. */
    fun executeSync(ctx: Long, operation: String, requestJson: String): String

    /** Start a background job; returns its handle (0 on rejection). */
    fun startJob(ctx: Long, operation: String, requestJson: String): Long

    /** Poll a job: the envelope JSON if finished, else `null`. */
    fun pollJob(job: Long): String?

    fun cancelJob(job: Long)

    fun freeJob(job: Long)
}
